### PR TITLE
Fix the external controller connection

### PR DESCRIPTION
**Pull request by ShuffleWire**
Was causing a loop eating 100% CPU.
Fix https://github.com/cyberbotics/webots/issues/3504

It is the correct way to fix it ? It seems to work, but I don't have enough knowledge of the code to say...
I've just noticed that the behavior between `void WbController::start()` (and the print "waiting for connection...") and `void WbController::disconnected()` (and the print "disconnected, waiting for new connection.") was different regarding this issue.

Hence, I've duplicate the code from `disconnected()`, which gave the correct behavior to `start()`.

### CODE REVIEW

**Comment by ygoumaz** [comment]
The fix doesn't seem to correct the bug you reported [in this comment](https://github.com/cyberbotics/webots/issues/3504#issuecomment-1276227746), namely:
- Start simulation with extern controller in realtime -> 0% CPU usage
- Activate fast mode -> 100% CPU usage 

Is this the bug you are trying to fix? As you reported this is only true before the controller starts. Once started, and if killed, the CPU goes back to 0% as it should.

**Comment by ygoumaz** [comment]
I tested a little more, and found an additional interesting element. The bug can be extended to the following scenario.
- Start Webots in realtime mode with an extern controller no yet connected.
- Activate fast mode (still no controller) ->100% CPU usage (bug).
- Connect the controller.
- Pause the simulation -> 0% CPU usage (correct behavior)
- Kill the controller -> 0% CPU usage (correct behavior)
- Without quitting Webots, put the simulation back to realtime mode -> still 0% CPU usage (correct behavior)
- Now activate the fast mode again. The CPU goes back to 100% even if no controller is connected (bug)

It seems to be the switch between realtime and fast modes when waiting for an extern controller that is the source of the problem.

**Comment by ShuffleWire** [comment]
Well ! 
You right, my fix didn't work... I don't know what test I've made, but that obviously don't work !
So "Is this the bug you are trying to fix? " Yes.
Your last comment is correct, however I'm not sure about the conclusion. To me it's look like webots is actively looking for a controller, hence eat 100% cpu.
The thing is that it does that in real mode as well, but, because the loop is slow down using the QTimer to maintain real time speed, it's not actually noticeable.

**Comment by ygoumaz** [comment]
I will also investigate to find if we can remove this CPU load.

**Comment by ygoumaz** [comment]
The source of the issue is the following:
- The controllers lists are not making any difference between connected and unconnected extern controllers. 
- Because of that, pressing the play button is moving the extern controller to the "started" list (called `mControllers`).
- The controller is started and `pauseStepTimer()` is called to pause the timer.
- Pressing the fast mode button is restarting the timer in the background, but because the controller is already started, the `pauseStepTimer()` method is never called, causing the `step()` function to be called in loop and use 100% CPU usage.

The proposed solution is the following:
The `mExternControllers` list (which was listing all unstarted extern controllers) is replaced by the `mDisconnectedExternControllers` list which tracks all extern controllers which are waiting for a connection. This way, if a step is executed and `mDisconnectedExternControllers` contains a controller, the timer is paused. 
The TCP server can also benefit from this cleanup.

File: src/webots/control/WbControlledWorld.hpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ class WbControlledWorld : public WbSimulationWorld {
   void step() override;
 
   QList<WbController *> controllers() const { return mControllers; }
-  QList<WbController *> externControllers() const { return mExternControllers; }
+  QList<WbController *> externControllers() const { return mDisconnectedExternControllers; }
```
Comment (omichel) [inline src/webots/control/WbControlledWorld.hpp:45]: Shall we also rename this method?
```suggestion
  QList<WbController *> disconnectedExternControllers() const { return mDisconnectedExternControllers; }
```

Comment (ygoumaz) [reply src/webots/control/WbControlledWorld.hpp:45]: I think it is a good idea, it will be more explicit.

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**